File: Problems/0399. (M) Evaluate Division/Python3/solution.py
class Solution:
  def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
    # iterative BFS
    
    adj = {}
    for i, edge in enumerate(equations):
      if edge[0] not in adj:
        adj[edge[0]] = []
      if edge[1] not in adj:
        adj[edge[1]] = []
      adj[edge[0]].append((edge[1], values[i]))
      adj[edge[1]].append((edge[0], 1/values[i]))
    
    ret = []
    for u, v in queries:
      ans = -1
      if u in adj and v in adj: 
        nodes = deque()
        nodes.append((u, 1))
        visited = set()
        while nodes:
          cur, f = nodes.popleft()
          if cur in visited or cur not in adj:
            continue
          if cur == v:
            ans = f
            break
          visited.add(cur)
          for nxt, n_f in adj[cur]:
            nodes.append((nxt, f*n_f))
      ret.append(ans)
    
    return ret

File: Problems/0399. (M) Evaluate Division/Java/solution.java
class Solution {
  public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
    // iterative BFS

    Map<String,List<Pair<String,Double>>> adj = new HashMap();
    for (int i = 0; i < equations.size(); i++) {
      if (!adj.containsKey(equations.get(i).get(0)))
        adj.put(equations.get(i).get(0), new ArrayList());
      if (!adj.containsKey(equations.get(i).get(1)))
        adj.put(equations.get(i).get(1), new ArrayList());
      adj.get(equations.get(i).get(0)).add(new Pair(equations.get(i).get(1), values[i]));
      adj.get(equations.get(i).get(1)).add(new Pair(equations.get(i).get(0), 1/values[i]));
    }

    double[] ret = new double[queries.size()];
    for (int i = 0; i < queries.size(); i++) {
      List<String> q = queries.get(i);
      double ans = -1;
      if (adj.containsKey(q.get(0)) && adj.containsKey(q.get(1))) {
        Deque<Pair<String,Double>> nodes = new ArrayDeque();
        nodes.add(new Pair(q.get(0), 1.0));
        Set<String> visited = new HashSet();
        while (nodes.size() > 0) {
          Pair<String,Double> cur = nodes.pollFirst();
          if (visited.contains(cur.getKey()) || !adj.containsKey(cur.getKey()))
            continue;
          if (cur.getKey().equals(q.get(1))) {
            ans = cur.getValue();
            break;
          }
          visited.add(cur.getKey());
          for (int j = 0; j < adj.get(cur.getKey()).size(); j++) {
            List<Pair<String,Double>> nxt = adj.get(cur.getKey());
            nodes.add(new Pair(nxt.get(j).getKey(), cur.getValue() * nxt.get(j).getValue()));
          }
        }
      }
      ret[i] = ans;
    }
    return ret;
  }
}
